### DISCUSSION TITLE
Underlying thread implementation

### DISCUSSION THREAD

**Discussion by istathar** [aesiniath/unbeliever#67] [Internals]
Right now **core-program**'s `forkThread` uses **async** ... because everyone thinks it's great and what you should use. Yup.

During a discussion with some friends at FP Complete, @snoyberg pointed out that in **unliftio** they switched to using GHC.Conc [^1] (?) directly rather than working through **async** and its use of Control.Concurrent getting around an implementation that was O(n*m) rather than O(n+1).

We really ought to investigate that and see if it's at all something we need to consider. Since we _do_ have a wrapper in `forkThread`, `waitThread` etc changing internals would be easy.

[^1]: https://hackage.haskell.org/package/base-4.15.0.0/docs/GHC-Conc.html

**Comment by snoyberg** [comment]
Just to clarify: I'm referring to the [`Conc`](https://www.stackage.org/haddock/nightly-2021-10-07/unliftio-0.2.20/UnliftIO-Async.html#t:Conc) data type. The advantages it provides are documented in this blog post: https://www.fpcomplete.com/blog/transformations-on-applicative-concurrent-computations/.
